dali/pipeline/data/tensor_vector.cc

```diff
@@ -172,240 +206,299 @@ void CopyImpl(DstBatch<DstBackend> &dst, const SrcBatch<SrcBackend> &src, const
+                           this->type(), ", new: ", type, error_suffix));
+
+  DALI_ENFORCE(this->sample_dim() == sample_dim,
+               make_string("Sample must have the same sample dim as the target batch, current: ",
+                           this->sample_dim(), ", new: ", sample_dim, error_suffix));
+
+  DALI_ENFORCE(this->GetLayout() == layout || layout.empty(),
+               make_string("Sample must have the same layout as the target batch current: ",
+                           this->GetLayout(), ", new: ", layout, " or come with empty layout ",
+                           error_suffix));
+
+  DALI_ENFORCE(this->is_pinned() == pinned,
+               make_string("Sample must have the same pinned status as target batch, current: ",
+                           this->is_pinned(), ", new: ", pinned, error_suffix));
+
+  DALI_ENFORCE(this->order() == order,
+               make_string("Sample must have the same order as the target batch", error_suffix));
+
+  DALI_ENFORCE(this->device_id() == device_id,
+               make_string("Sample must have the same device id as target batch, current: ",
+                           this->device_id(), ", new: ", device_id, error_suffix));
+}
+
+
 template <typename Backend>
 void TensorVector<Backend>::UnsafeSetSample(int sample_idx, const TensorVector<Backend> &src,
                                             int src_sample_idx) {
-  // TODO(klecki): more consistency checks, contiguous -> non-contiguous removes shares_data from
-  // samples
   // Bounds check
   assert(sample_idx >= 0 && sample_idx < curr_num_tensors_);
   assert(src_sample_idx >= 0 && src_sample_idx < src.curr_num_tensors_);
-  DALI_ENFORCE(type() == src.type(),
-               make_string("Sample must have the same type as a target batch, current: ", type(),
-                           " new: ", src.type(), " for ", sample_idx, " <- ", src_sample_idx, "."));
-  DALI_ENFORCE(sample_dim() == src.shape().sample_dim(),
-               make_string("Sample must have the same dimensionality as a target batch, current: ",
-                           sample_dim(), " new: ", src.shape().sample_dim(), " for ", sample_idx,
-                           " <- ", src_sample_idx, "."));
-  DALI_ENFORCE(this->order() == src.order(), "Sample must have the same order as a target batch");
-  DALI_ENFORCE(
-      GetLayout() == src.GetLayout(),
-      make_string("Sample must have the same layout as a target batch current: ", GetLayout(),
-                  " new: ", src.GetLayout(), " for ", sample_idx, " <- ", src_sample_idx, "."));
-  DALI_ENFORCE(
-      is_pinned() == src.is_pinned(),
-      make_string("Sample must have the same pinned status as target batch, current: ", is_pinned(),
-                  " new: ", src.is_pinned(), " for ", sample_idx, " <- ", src_sample_idx, "."));
+  // Setting any individual sample converts the batch to non-contiguous mode
+  MakeNoncontiguous();
```

Comment:
Hmm, that's the question of semantics and function invariants. I am less surprised when setting an individual sample converts the batch into noncontiguous mode (so individual samples), rather than getting lucky and keeping the contiguous status.